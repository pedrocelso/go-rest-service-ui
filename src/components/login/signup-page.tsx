import React, { Component } from 'react';
import { connect } from 'react-redux'
import { toast } from 'react-toastify';
import { equals } from 'ramda';
import { LinearProgress, Typography, Grid, Button } from '@material-ui/core';
import { Link } from 'react-router-dom'
import { isValidEmail } from './login-page'
import './login-page.scss'
import { authenticate } from './login-actions'
import { UserService } from '../../services/user'

interface SignUpPageProps {
  authenticate: typeof authenticate
  service: UserService;
}

interface SignUpPageState {
  email: string
  name: string
  password: string
  passwordConfirmation: string
  loading: boolean
}

const notify = (msg: string, feeling: number) => {
  switch (feeling) {
    case (-1):
      toast.error(msg);
  }
}

export class SignUpPage extends Component<SignUpPageProps, SignUpPageState> {
  constructor(props: SignUpPageProps) {
    super(props)
    this.state = {
      email: ``,
      name: ``,
      password: ``,
      passwordConfirmation: ``,
      loading: false
    }
  }

  handleEmailChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget
    this.setState({ email: value })
  }

  handlePasswordChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget
    this.setState({ password: value })
  }

  handlePasswordConfirmationChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget
    this.setState({ passwordConfirmation: value })
  }

  handleNameChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget
    this.setState({ name: value })
  }

  validateForm = () => {
    const { email, password, passwordConfirmation, name } = this.state

    if (!isValidEmail(email)) {
      return false
    }

    if (!name || !password) {
      return false
    }

    if (!equals(password, passwordConfirmation)) {
      return false
    }

    return true
  }

  signUp = () => {
    const { service } = this.props
    const { email, password, passwordConfirmation, name } = this.state

    if (isValidEmail(email) && !!name && equals(password, passwordConfirmation)) {
      this.setState({ loading: true })
      service.authenticate({ email, password })
        .fork(
          (e) => {
            this.setState({ loading: false })
            if (e.statusCode === 401) {
              notify(`Wrong email/password!`, -1)
            }
          },
          ({ token }) => {
            this.setState({ loading: false })
            sessionStorage.setItem(`jwtToken`, token)
            authenticate(token)
          }
        )
    } else {
      notify(`Invalid credentials!`, -1)
    }
  }

  render() {
    const { email, loading } = this.state

    const validData = this.validateForm()

    const loadingbar = loading ? <LinearProgress variant="indeterminate" /> : null

    return (
      <div className="modal__dialog">
        <Grid container className="modal__grid" alignContent="center" spacing={16} direction="column">
          {loadingbar}
          <Grid item xs={10} container alignContent="center" direction="column">
            <Typography variant="h4">Sign Up</Typography>
          </Grid>
          <Grid item xs={10} container direction="row">
            <input type="text" className="form-control" name="name" placeholder="Name" onChange={this.handleNameChange} />
          </Grid>
          <Grid item xs={10} container direction="row">
            <input type="text" className={`form-control ${!isValidEmail(email) ? "is-invalid" : ""}`} name="email" placeholder="E-mail" onChange={this.handleEmailChange} />
          </Grid>
          <Grid item xs={10} container direction="row">
            <input type="password" className="form-control" name="password" placeholder="Password" onChange={this.handlePasswordChange} />
          </Grid>
          <Grid item xs={10} container direction="row">
            <input type="password" className="form-control" name="passwordConfirmation" placeholder="Confirm Password" onChange={this.handlePasswordConfirmationChange} />
          </Grid>
          <Grid item xs={10} container direction="column">
            <Button disabled={!validData} variant="contained" color="primary" onClick={this.signUp} className="btnSignUp">
              Sign Up
            </Button>
          </Grid>
          <Grid item xs={10} container alignContent="center" direction="column">
            <Typography>Don't have an account? <Link to="/signup">Click here</Link></Typography>
          </Grid>
        </Grid>
      </div>
    )
  }
}

const mapDispatchToProps = { authenticate }
export default connect(null, mapDispatchToProps)(SignUpPage)