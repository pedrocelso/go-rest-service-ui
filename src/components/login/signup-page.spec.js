import React from 'react';
import { shallow } from 'enzyme'
import sinon from 'sinon'

import { SignUpPage } from './signup-page'

const baseProps = {
  authenticate: sinon.fake(),
  service: {
    authenticate: (_) => ({
      fork: (rej, res) => {
        res({ token: `token!` })
      }
    })
  }
}

describe(`<SignUpPage />`, () => {
  it(`Should render without errors`, () => {
    const component = shallow(
      <SignUpPage {...baseProps} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`.handlePasswordChange() - Setting valid name`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.instance().handlePasswordChange({ currentTarget: { value: `bororoska` } })
    expect(wrapper.state().password).toEqual(`bororoska`)
  });

  it(`.handleEmailChange() - Setting valid email`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<SignUpPage {...props} />);
    wrapper.instance().handleEmailChange({ currentTarget: { value: `bororoska@test.com` } })
    expect(wrapper.state().email).toEqual(`bororoska@test.com`)
  });

  it(`.handlePasswordConfirmationChange() - Setting valid password confirmation`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<SignUpPage {...props} />);
    wrapper.instance().handlePasswordConfirmationChange({ currentTarget: { value: `bororoska` } })
    expect(wrapper.state().passwordConfirmation).toEqual(`bororoska`)
  });

  it(`.handleNameChange() - Setting valid name`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<SignUpPage {...props} />);
    wrapper.instance().handleNameChange({ currentTarget: { value: `Agric Malmim` } })
    expect(wrapper.state().name).toEqual(`Agric Malmim`)
  });

  it(`.signUp() - valid state`, () => {
    // const authenticate = sinon.spy()

    // const props = {
    //   ...baseProps,
    //   authenticate
    // }

    // const wrapper = shallow(<SignUpPage {...props} />);

    // wrapper.setState({ password: 'bar', email: 'foo@bar.com' });

    // wrapper.instance().login();
    // expect(authenticate.called).toBeTruthy();
  });

  it(`.validateForm() - valid state`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foo@bar.com', name: `Foo`, passwordConfirmation: `bar` });

    expect(wrapper.instance().validateForm()).toBeTruthy();
  });

  it(`.validateForm() - invalid email`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foobar.com', name: `Foo`, passwordConfirmation: `bar` });

    expect(wrapper.instance().validateForm()).toBeFalsy();
  });

  it(`.validateForm() - empty name`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foo@bar.com', passwordConfirmation: `bar` });

    expect(wrapper.instance().validateForm()).toBeFalsy();
  });

  it(`.validateForm() - empty password`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.setState({ password: '', email: 'foo@bar.com', name: `Foo`, passwordConfirmation: `bar` });

    expect(wrapper.instance().validateForm()).toBeFalsy();
  });

  it(`.validateForm() - password not matching confirmation`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<SignUpPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foo@bar.com', name: `Foo`, passwordConfirmation: `bar1` });

    expect(wrapper.instance().validateForm()).toBeFalsy();
  });
});