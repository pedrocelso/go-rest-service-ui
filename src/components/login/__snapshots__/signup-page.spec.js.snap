// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUpPage /> Should render without errors 1`] = `
<div>
  <div
    className="modal__dialog"
  >
    <WithStyles(Grid)
      alignContent="center"
      className="modal__grid"
      container={true}
      direction="column"
      spacing={16}
    >
      <WithStyles(Grid)
        alignContent="center"
        container={true}
        direction="column"
        item={true}
        xs={10}
      >
        <WithStyles(Typography)
          variant="h4"
        >
          Sign Up
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        container={true}
        direction="row"
        item={true}
        xs={10}
      >
        <input
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        container={true}
        direction="row"
        item={true}
        xs={10}
      >
        <input
          className="form-control is-invalid"
          name="email"
          onChange={[Function]}
          placeholder="E-mail"
          type="text"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        container={true}
        direction="row"
        item={true}
        xs={10}
      >
        <input
          className="form-control"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        container={true}
        direction="row"
        item={true}
        xs={10}
      >
        <input
          className="form-control"
          name="passwordConfirmation"
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        container={true}
        direction="column"
        item={true}
        xs={10}
      >
        <WithStyles(Button)
          className="btnSignUp"
          color="primary"
          disabled={true}
          onClick={[Function]}
          variant="contained"
        >
          Sign Up
        </WithStyles(Button)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        alignContent="center"
        container={true}
        direction="column"
        item={true}
        xs={10}
      >
        <WithStyles(Typography)>
          Already registered? 
          <Link
            to="/"
          >
            Click here
          </Link>
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
</div>
`;
